source/Playlist.cpp

```diff
@@ -0,0 +1,142 @@
+/* Playlist.cpp
+Copyright (c) 2022 by RisingLeaf
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Playlist.h"
+
+#include "DataNode.h"
+#include "GameData.h"
+#include "PlayerInfo.h"
+#include "System.h"
+
+#include <algorithm>
+
+using namespace std;
+
+namespace {
+	const Track *currentTrack = nullptr;
+}
+
+
+
+Playlist::Playlist(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Playlist::Load(const DataNode &node)
+{
+	if(!name.empty())
+	{
+		node.PrintTrace("Error: Duplicate definition of playlist:");
+		return;
+	}
+	name = node.Token(1);
+
+	for(const DataNode &child : node)
+	{
+		const string &key = child.Token(0);
+		bool hasValue = child.Size() >= 2;
+		if(key == "to" && child.Token(1) == "play")
+			toPlay.Load(child);
+		else if(key == "location")
+			location.Load(child);
+		else if(key == "priority" && hasValue)
+			priority = max<unsigned>(0, child.Value(1));
+		else if(key == "weight" && hasValue)
+			weight = max<unsigned>(1, child.Value(1));
```

Comment:
These should print warnings if the user-defined values get ignored, as it's not intuitive that a weight of 0.5 doesn't function, or that a negative priority doesn't get less priority than a default-valued playlist.

Are these even necessary? I think checking for a positive weight should be enough.

Reply:
now checking for negative values -> warning